File: scripts/utils/formatter.go
package main

import (
	"fmt"
	"log"
	"os"
	"regexp"
	"strings"
)

func extractSectionFromFile(filePath, startMarker, endMarker string) (string, error) {
	data, err := os.ReadFile(filePath)
	if err != nil {
		return "", fmt.Errorf("error reading file: %v", err)
	}
	text := string(data)

	startIndex := strings.Index(text, startMarker)
	if startIndex == -1 {
		return "", fmt.Errorf("start marker %q not found", startMarker)
	}
	startIndex += len(startMarker)

	endIndex := strings.Index(text, endMarker)
	if endIndex == -1 {
		return "", fmt.Errorf("end marker %q not found", endMarker)
	}

	backendSection := text[startIndex:endIndex]
	return backendSection, nil
}

func extractAfterStart(filePath, startMarker string) (string, error) {
	data, err := os.ReadFile(filePath)
	if err != nil {
		return "", fmt.Errorf("error reading file: %v", err)
	}
	text := string(data)

	startIndex := strings.Index(text, startMarker)
	if startIndex == -1 {
		return "", fmt.Errorf("start marker %q not found", startMarker)
	}
	startIndex += len(startMarker)

	endIndex := len(text)-1

	backendSection := text[startIndex:endIndex]
	return backendSection, nil
}

func main() {
	backendFileName := "RELEASE.md"
	backendStartMarker := "# Jaeger Backend Release Process"
	backendEndMarker := "## Patch Release"
	backendSection, err := extractSectionFromFile(backendFileName, backendStartMarker, backendEndMarker)
	if err != nil {
		log.Fatalf("Failed to extract backendSection: %v", err)
	}
	re_star := regexp.MustCompile(`(\n\s*)(\*)(\s)`)
	backendSection = re_star.ReplaceAllString(backendSection, "$1$2 [ ]$3")
	re_num :=regexp.MustCompile(`(\n\s*)([0-9]*\.)(\s)`)
	backendSection = re_num.ReplaceAllString(backendSection, "$1* [ ]$3")

	docFilename:= "DOC_RELEASE.md"
	docStartMarker := "# Release instructions"
	docEndMarker := "### Auto-generated documentation for CLI flags"
	docSection, err := extractSectionFromFile(docFilename, docStartMarker, docEndMarker)
	if err != nil{
		log.Fatalf("Failed to extract documentation section: %v", err)
		
	}
	re_dash :=regexp.MustCompile(`(\n\s*)(\-)`)
	docSection=re_dash.ReplaceAllString(docSection, "$1* [ ]")
	
	uiFilename := "jaeger-ui/RELEASE.md"
	uiStartMarker := "# Cutting a Jaeger UI release"
	uiSection, err := extractAfterStart(uiFilename, uiStartMarker)
	if err != nil{
		log.Fatalf("Failed to extract UI section: %v", err)
	}
	uiSection = re_dash.ReplaceAllString(uiSection, "$1$2 [ ]$3")
	uiSection = re_num.ReplaceAllString(uiSection, "$1* [ ]$3")
	


	fmt.Println("# UI Release")
	fmt.Println(uiSection)
	fmt.Println("# Backend Release")
	fmt.Println(backendSection)
	fmt.Println("# Doc Release")
	fmt.Println(docSection)

}

File: scripts/utils/formatter.py
#!/usr/bin/env python3
# Copyright (c) 2025 The Jaeger Authors.
# SPDX-License-Identifier: Apache-2.0

import re
import sys

def extract_section_from_file(file_path, start_marker, end_marker):
    with open(file_path, 'r') as f:
        text = f.read()
    start_index = text.find(start_marker)
    if start_index == -1:
        raise Exception(f"start marker {start_marker!r} not found")
    start_index += len(start_marker)
    end_index = text.find(end_marker)
    if end_index == -1:
        raise Exception(f"end marker {end_marker!r} not found")
    return text[start_index:end_index]

def replace_star(text):
    re_star = re.compile(r'(\n\s*)(\*)(\s)')
    text = re_star.sub(r'\1\2 [ ]\3', text)
    return text

def replace_dash(text):
    re_dash = re.compile(r'(\n\s*)(\-)')
    text = re_dash.sub(r'\1* [ ]', text)
    return text

def replace_num(text):
    re_num = re.compile(r'(\n\s*)([0-9]*\.)(\s)')
    text = re_num.sub(r'\1* [ ]\3', text)
    return text

def replace_version(ui_text, backend_text, doc_text, pattern, ver):
    ui_text = re.sub(pattern, ver, ui_text)
    backend_text = re.sub(pattern, ver, backend_text)
    doc_text = re.sub(pattern, ver, doc_text)
    return ui_text, backend_text, doc_text

def fetch_content(file_name):
    start_marker = "<!-- BEGIN_CHECKLIST -->"
    end_marker = "<!-- END_CHECKLIST -->"
    text = extract_section_from_file(file_name, start_marker, end_marker)
    return text

def main():
    
    loc = sys.argv[1]
    v1 = sys.argv[2]
    v2 = sys.argv[3]
    try:
        backend_file_name = "RELEASE.md"
        backend_section = fetch_content(backend_file_name)
    except Exception as e:
        sys.exit(f"Failed to extract backendSection: {e}")
    backend_section = replace_star(backend_section)
    backend_section = replace_num(backend_section)
    try:
        doc_filename = loc
        doc_section = fetch_content(doc_filename)
    except Exception as e:
        sys.exit(f"Failed to extract documentation section: {e}")
    doc_section=replace_dash(doc_section)

    try:
        ui_filename = "jaeger-ui/RELEASE.md"
        ui_section = fetch_content(ui_filename)
    except Exception as e:
        sys.exit(f"Failed to extract UI section: {e}")

    ui_section=replace_dash(ui_section)
    ui_section=replace_num(ui_section)

    #Concrete version
    v1_pattern = r'(?:X\.Y\.Z|1\.[0-9]+\.[0-9]+|1\.x\.x)'
    ui_section, backend_section, doc_section = replace_version(ui_section, backend_section, doc_section, v1_pattern, v1)
    v2_pattern = r'2.x.x'
    ui_section, backend_section, doc_section = replace_version(ui_section, backend_section, doc_section, v2_pattern, v2)

    print("# UI Release")
    print(ui_section)
    print("# Backend Release")
    print(backend_section)
    print("# Doc Release")
    print(doc_section)

if __name__ == "__main__":
    main()
